### DISCUSSION TITLE
How can i accessing the selected record from bulk action in filament table and access the records from form method?

### DISCUSSION THREAD

**Discussion by christmex** [filamentphp/filament#8320] [Help] [answered]
### **What i want?**
Accessing the selected record from bulk action in filament table and accessing the records from the form method

### **Explanation of what i want to achieve**
When I click my bulk action, I open a new form in modal where the user can select which classroom the student will promoted based on the student's active classroom, so I need to get the student records to check they are in which class now, i hope it is clear 🙏

### **What did I do?**
Here's my code sneak peek
```php
Tables\Actions\BulkAction::make('promoted_students')
  ->form([
        Forms\Components\Select::make('classroom_id')
              ->live()
              //Here's what I did to try to accessing the selected records, but it doesn't work
              ->options(function(Collection $records){
                  return Classroom::all()
                  ->pluck('classroom_name', 'id');
              })
              ->searchable(['classroom_name'])
              ->default(fn($state) => $state)
              ->required(),
]),
```

Where did I miss? thank you for letting me know, thank you for any helps 🙏

**Comment by danharrin** [comment, marked as answer]
Try `$livewire->getSelectedTableRecords()`

**Reply by christmex** [reply]
Thank you for the answer 🙏

**Reply by dr-codswallop** [reply]
Would you mind sharing how you implemented `$livewire->getSelectedTableRecords()` 🙏

I'm facing the exact situation where I'd need to limit options to only those selected for the bulk action.

**Reply by denizaygundev** [reply]
@dr-codswallop in filament3 I actually found that it turns out the calculated record data is available on the model on `$records` anyway 
```
                ->action(function (array $data, Collection $records, $table, $livewire): void {
                        ray($livewire);
                        ray($livewire->getSelectedTableRecords());
                        ray($table);
                        ```

**Comment by christmex** [comment]
@dr-codswallop 

hi Mate, here is the code, I hope you found it helpful, let us know if you found any issue

```php

Tables\Actions\BulkAction::make('buln_name')
    ->form([
        \Filament\Forms\Components\Select::make('column_name')
            ->required(),
            //.... another field here
    ])
    ->action(function(array $data, $livewire){
        // dd($livewire->selectedTableRecords, $data);
        // $data contains the form value, and the $livewire contains many stuff, but for getting only the selected data you can use $livewire->selectedTableRecords 

    })
```

**Comment by dr-codswallop** [comment]
Cheers @christmex, so have I got this correct? 

Say in my table of languages I choose multiple records for bulk actioning. I want to have only those selected as options (either in Radio, Select, CheckboxList) in the modal form. And would I use the exact same method if I wanted to set the default? 

The following seems to work, just unsure if it's the advised method. 

```php
Tables\Actions\BulkAction::make('merge_records')
    ->form([
        \Filament\Forms\Components\Radio::make('master_id')
            ->label('Choose a master record to merge all selected records into')
            ->options(function($livewire) {
                $options = $livewire->getSelectedTableRecords();
                return $options->pluck('language', 'id'); // whatever columns are necessary
            })
            ->default(function($livewire) {
                $options = $livewire->getSelectedTableRecords();
                return $options->first()->id; // if I wanted to have the first item pre-selected
            })
            ->required(),
    ])
    ->action(function(array $data, $livewire){
        // dd($livewire->selectedTableRecords, $data);
        // $data contains the form value, and the $livewire contains many stuff, but for getting only the selected data you can use $livewire->selectedTableRecords 
        // DO THE MERGING STUFF
    })
```

**Reply by christmex** [reply]
Well it looks good man @dr-codswallop

**Comment by photojo** [comment]
How can I achieve, that each selected row is listet when selected, f.e. in a sidebar, without click to BulkAction-Button?

**Reply by christmex** [reply]
hi @photojo so u mean u want to create clickable row? when u clik the row it will select the row?

**Reply by photojo** [reply]
Hi @christmex : no, I don’t want clickable rows. I just wanted a possibility so show selected rows, before I choose a bulk-action.

**Comment by almighty-lilith** [comment]
$livewire->getSelectedTableRecords() returns an empty array on my side...

Illuminate\Database\Eloquent\Collection {#2182 ▼ // app\Filament\Resources\BookAdminResource\Pages\ListBookAdmins.php:32
  #items: []
  #escapeWhenCastingToString: false
}

any idea why ?

**Reply by christmex** [reply]
@almighty-lilith can we see your code?

**Reply by almighty-lilith** [reply]
The proper action : 

\app\Filament\Resources\BookAdminResource\Pages\ListBookAdmins.php
`

class ListBookAdmins extends ListRecords

{
    protected static string $resource = BookAdminResource::class;



    public static function bulkAddTagsAction(): Action
    {
        return Action::make('bulkAddTagsAction')
            ->label('Ajouter un mot-clé')
            ->modalDescription('Ajouter un ou plusieurs mot-clés à tous les livres sélectionnés')
            ->icon('heroicon-o-tag')
            ->form([
                Forms\Components\Select::make('add_tags')
                    ->label('Mots-clés')
                    ->multiple()
                    ->relationship('tags', 'title')
                    ->dehydrated()  // pour récupérer les données dans la fonction action à partir d"une relation
                    ->preload(),

            ])
            ->action(function (array $data, $livewire): void {

                dd($data, $livewire->getSelectedTableRecords());`
                
                
The bulkactions in the main resource :

            ->bulkActions([
                Tables\Actions\BulkActionGroup::make([
                    Tables\Actions\DeleteBulkAction::make(),
                    Pages\ListBookAdmins::bulkAddTagsAction(),
                ]),

            ])
            
            
            
The goal is to attach, one or several tags to several books through the selected bulk items -> model -> tags selected
The attachment of several tags is working well in a book edit form field

![the dump in the add_tags modal](https://github.com/user-attachments/assets/f889ba9c-1667-42ba-abb3-255effd573d3)
![image](https://github.com/user-attachments/assets/bc711521-c191-4089-b152-12c0e38f3426)
![image](https://github.com/user-attachments/assets/1f24b775-a3ed-4f9b-8027-a27d01781311)

Thank you for your time !

**Reply by monteduro** [reply]
What I would like to do is enable or disable the visibility of the bulk action based on the selected records.

For example, if I select two records that are in draft, I want to display only the "Publish" button. On the other hand, if I select records that are already published, I want to show only the "Move to Draft" option in the bulk actions, and so on.

Apparently, if I use `getSelectedTableRecords()` inside` visible()`, it is only evaluated on page load and not when selecting rows dynamically.
